File: app/src/main/java/com/yuta/__old/projects/dialog/DeleteProjectDialog.java
package com.yuta.__old.projects.dialog;

import android.content.Context;
import android.widget.TextView;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import com.yuta.__old.R;
import com.yuta.__old.projects.ProjectsFragment;
import com.yuta.app.domain.model.entity.ProjectDto;
import com.yuta.app.domain.model.response.UpdateResponse;
import com.yuta.common.ui.CancelableDialog;
import com.yuta.app.network.RequestViewModel;

public class DeleteProjectDialog extends CancelableDialog {
    private final ProjectDto project;
    private RequestViewModel viewModel;

    public DeleteProjectDialog(Context context, Fragment fragment, ProjectDto project) {
        super(context, fragment);
        setDialogLayout(R.layout.dialog_delete);
        this.project = project;
    }

    @Override
    public void start() {
        super.start();
        viewModel = new ViewModelProvider(fragment).get(RequestViewModel.class);
        setupTextView(project.getName());

        dialog.findViewById(R.id.close).setOnClickListener(v -> dismiss());
        dialog.findViewById(R.id.submit).setOnClickListener(v -> {
            deleteProject(fragment, project);
            dismiss();
        });
    }

    private void setupTextView(String name) {
        String text = getContext().getString(R.string.delete_project_desc) + " \"" + name + "\"?";
        ((TextView) dialog.findViewById(R.id.name_desc)).setText(text);
    }

    private void deleteProject(Fragment fragment, ProjectDto project) {
        viewModel.getResultLiveData().removeObservers(fragment);
        viewModel.deleteProject(project.getId());
        viewModel.getResultLiveData().observe(fragment, result -> {
            if (!(result instanceof UpdateResponse)) return;
            ((ProjectsFragment) fragment).updateLists();
            dismiss();
        });
    }
}

File: app/src/main/java/com/yuta/projects/ui/dialog/DeleteProjectDialog.kt
package com.yuta.projects.ui.dialog

import android.widget.Button
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.viewModelScope
import com.yuta.app.R
import com.yuta.common.ui.CancelableDialog
import com.yuta.domain.model.ProjectDto
import com.yuta.projects.viewmodel.ProjectDialogsViewModel
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlin.getValue

class DeleteProjectDialog(
    private val project: ProjectDto,
    fragment: Fragment,
    private val onDeleteSuccessCallback: () -> Unit
) : CancelableDialog(R.layout.dialog_delete, fragment.requireActivity()) {

    private val submitButton: Button by lazy { dialog.findViewById(R.id.submit) }
    private val closeButton: Button by lazy { dialog.findViewById(R.id.close) }
    private val description: TextView by lazy { dialog.findViewById(R.id.desc) }

    private val projectViewModel: ProjectDialogsViewModel by fragment.viewModels()

    override fun start() {
        super.start()
        setupTextView(project.name)

        closeButton.setOnClickListener { dismiss() }
        submitButton.setOnClickListener { deleteProject(project.id) }
    }

    private fun setupTextView(name: String) {
        val text = "${context.getString(R.string.delete_project_desc)} \"$name\"?"
        description.text = text
    }

    private fun deleteProject(id: Int) {
        projectViewModel.viewModelScope.launch {
            val isDeleted = projectViewModel.delete(id).first()
            handleDeleteResult(isDeleted)
        }
    }

    private fun handleDeleteResult(isDeleted: Boolean) {
        if (isDeleted) {
            onDeleteSuccessCallback()
            dismiss()
        }
    }
}
